src/sage/rings/function_field/drinfeld_modules/morphism.py

```diff
@@ -404,3 +419,365 @@ def ore_polynomial(self):
+              Defn: t^6
+
+        """
+        return H(self.ore_polynomial() * other.ore_polynomial())
+
+    def __invert__(self):
+        r"""
+        Return the inverse of this morphism.
+
+        Only morphisms defined by constant nonzero Ore
+        polynomials are invertible.
+
+        EXAMPLES::
+
+            sage: Fq = GF(5)
+            sage: A.<T> = Fq[]
+            sage: K.<z> = Fq.extension(3)
+            sage: phi = DrinfeldModule(A, [z, 1, z, z^2])
+            sage: f = phi.hom(2); f
+            Endomorphism of Drinfeld module defined by T |--> z^2*t^3 + z*t^2 + t + z
+              Defn: 2
+            sage: f.inverse()  # indirect doctest
+            Endomorphism of Drinfeld module defined by T |--> z^2*t^3 + z*t^2 + t + z
+              Defn: 3
+
+        ::
+
+            sage: F = phi.frobenius_endomorphism()
+            sage: F.inverse()  # indirect doctest
+            Traceback (most recent call last):
+            ...
+            ZeroDivisionError: this morphism is not invertible
+
+        """
+        if not self.is_isomorphism():
+            raise ZeroDivisionError("this morphism is not invertible")
+        return self.parent()(~(self.ore_polynomial()[0]))
+
+    def _motive_matrix(self):
+        r"""
+        Return the matrix giving the action of this morphism
+        on the motives of the underlying Drinfeld modules.
+
+        For internal use. Do not call this method directly.
+
+        EXAMPLES::
+
+            sage: Fq = GF(5)
+            sage: A.<T> = Fq[]
+            sage: K.<z> = Fq.extension(3)
+            sage: phi = DrinfeldModule(A, [z, 0, 1])
+            sage: t = phi.ore_variable()
+            sage: u = t^2 + (2*z^2 + 3*z + 3)*t + (2*z + 3)
+            sage: f = phi.hom(u)
+            sage: f._motive_matrix()
+            [                      T + 3 + z                 3 + 3*z + 2*z^2]
+            [(1 + z + z^2)*T + 3 + 2*z - z^2               T + 2 - z + 2*z^2]
+
+        """
+        phi = self.domain()
+        phiT = phi.gen()
+        r = phiT.degree()
+        K = phi.base_over_constants_field()
+        S = phi.ore_polring()
+        Frob = S.twisting_morphism()
+        KT = PolynomialRing(K, name='T')
+
+        # The first row:
+        # we write u = u0 + u1*phiT + u2*phiT^2 + ...
+        u = self.ore_polynomial()
+        us = [ ]
+        while not u.is_zero():
+            u, ui = u.right_quo_rem(phiT)
+            us.append(ui)
+        l = len(us)
+        row = [KT([us[i][j] for i in range(l)]) for j in range(r)]
+        rows = [row]
+
+        # The next rows:
+        # each row is obtained from the previous one by
+        # applying the semi-linear transformation f |-> t*f
+        inv = K(phiT[r]).inverse()
+        B = inv * phiT
+        T = KT.gen()
+        for i in range(1, r):
+            row = [c.map_coefficients(Frob) for c in row]
+            row = [(inv*T - B[0]) * row[-1]] + [row[j-1] - B[j]*row[-1] for j in range(1, r)]
```

Comment:
pretty sure this is a long line but looks like it passed the linter so not sure what happened